File: src/sdk/python/hanami_sdk/hanami_sdk/task.py
# Copyright 2022 Tobias Anker
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from . import hanami_request
import json


def create_train_task(token: str,
                      address: str,
                      name: str,
                      cluster_uuid: str,
                      inputs: list,
                      outputs: list,
                      timeLength: int = 1,
                      verify_connection: bool = True) -> str:
    path = "/v1.0alpha/task/train"
    json_body = {
        "name": name,
        "cluster_uuid": cluster_uuid,
        "inputs": inputs,
        "outputs": outputs,
        "time_length": timeLength
    }
    body_str = json.dumps(json_body)
    return hanami_request.send_post_request(token,
                                            address,
                                            path,
                                            body_str,
                                            verify=verify_connection)


def create_request_task(token: str,
                        address: str,
                        name: str,
                        cluster_uuid: str,
                        inputs: list,
                        results: list,
                        timeLength: int = 1,
                        verify_connection: bool = True) -> str:
    path = "/v1.0alpha/task/request"
    json_body = {
        "name": name,
        "cluster_uuid": cluster_uuid,
        "inputs": inputs,
        "results": results,
        "time_length": timeLength
    }
    body_str = json.dumps(json_body)
    return hanami_request.send_post_request(token,
                                            address,
                                            path,
                                            body_str,
                                            verify=verify_connection)


def get_task(token: str,
             address: str,
             task_uuid: str,
             cluster_uuid: str,
             verify_connection: bool = True) -> str:
    path = "/v1.0alpha/task"
    values = f'uuid={task_uuid}&cluster_uuid={cluster_uuid}'
    return hanami_request.send_get_request(token,
                                           address,
                                           path,
                                           values,
                                           verify=verify_connection)


def list_tasks(token: str,
               address: str,
               cluster_uuid: str,
               verify_connection: bool = True) -> str:
    path = "/v1.0alpha/task/all"
    values = f'cluster_uuid={cluster_uuid}'
    return hanami_request.send_get_request(token,
                                           address,
                                           path,
                                           values,
                                           verify=verify_connection)


def delete_task(token: str,
                address: str,
                task_uuid: str,
                cluster_uuid: str,
                verify_connection: bool = True) -> str:
    path = "/v1.0alpha/task"
    values = f'uuid={task_uuid}&cluster_uuid={cluster_uuid}'
    return hanami_request.send_delete_request(token,
                                              address,
                                              path,
                                              values,
                                              verify=verify_connection)

File: src/sdk/go/ainari_sdk/task.go
/**
 * @author      Tobias Anker <tobias.anker@kitsunemimi.moe>
 *
 * @copyright   Apache License Version 2.0
 *
 *      Copyright 2022-2026 Tobias Anker <tobias.anker@kitsunemimi.moe>
 *
 *      Licensed under the Apache License, Version 2.0 (the "License");
 *      you may not use this file except in compliance with the License.
 *      You may obtain a copy of the License at
 *
 *          http://www.apache.org/licenses/LICENSE-2.0
 *
 *      Unless required by applicable law or agreed to in writing, software
 *      distributed under the License is distributed on an "AS IS" BASIS,
 *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *      See the License for the specific language governing permissions and
 *      limitations under the License.
 */

package ainari_sdk

import (
	"fmt"
)

type TaskInput struct {
	HexagonName        string `json:"hexagon"`
	DatasetColumnName  string `json:"dataset_column"`
	DatasetUuid        string `json:"dataset_uuid"`
}

type TaskResult struct {
	HexagonName        string `json:"hexagon"`
	DatasetColumnName  string `json:"dataset_column"`
}

func CreateTrainTask(context AccessContext, toriiPort int, name, instanceUuid string, inputs, outputs []TaskInput, number_of_epochs, timeLenght int) (map[string]interface{}, error) {
	address := fmt.Sprintf("%s:%d", context.ToriiBaseAddress, toriiPort)
    var inputArray []interface{}
    for _, input := range inputs {
        inputArray = append(inputArray, input)
    }

    var outputArray []interface{}
    for _, output := range outputs {
        outputArray = append(outputArray, output)
    }

	path := fmt.Sprintf("v1alpha/instance/%s/task/train", instanceUuid)
	jsonBody := map[string]interface{}{
		"name":             name,
		"number_of_epochs": number_of_epochs,
		"inputs":           inputArray,
		"outputs":          outputArray,
		"time_length":      timeLenght,
	}
	return SendPost(context, address, path, jsonBody)
}

func CreateRequestTask(context AccessContext, toriiPort int, name, instanceUuid string, inputs []TaskInput, results []TaskResult, timeLenght int) (map[string]interface{}, error) {
	address := fmt.Sprintf("%s:%d", context.ToriiBaseAddress, toriiPort)
	var inputArray []interface{}
    for _, input := range inputs {
        inputArray = append(inputArray, input)
    }

    var resultArray []interface{}
    for _, result := range results {
        resultArray = append(resultArray, result)
    }

	path := fmt.Sprintf("v1alpha/instance/%s/task/request", instanceUuid)
	jsonBody := map[string]interface{}{
		"name":         name,
		"inputs":       inputArray,
		"results":      resultArray,
		"time_length":  timeLenght,
	}
	return SendPost(context, address, path, jsonBody)
}

func CreateCheckpointSaveTask(context AccessContext, toriiPort int, name, instanceUuid string) (map[string]interface{}, error) {
	address := fmt.Sprintf("%s:%d", context.ToriiBaseAddress, toriiPort)
	path := fmt.Sprintf("v1alpha/instance/%s/task/checkpoint_save", instanceUuid)
	jsonBody := map[string]interface{}{
		"name": name,
	}
	return SendPost(context, address, path, jsonBody)
}

func CreateCheckpointRestoreTask(context AccessContext, toriiPort int, name, instanceUuid, checkpointUuid string) (map[string]interface{}, error) {
	address := fmt.Sprintf("%s:%d", context.ToriiBaseAddress, toriiPort)
	path := fmt.Sprintf("v1alpha/instance/%s/task/checkpoint_restore", instanceUuid)
	jsonBody := map[string]interface{}{
		"name": name,
		"checkpoint_uuid": checkpointUuid,
	}
	return SendPost(context, address, path, jsonBody)
}

func GetTask(context AccessContext, toriiPort int, taskUuid, instanceUuid string) (map[string]interface{}, error) {
	address := fmt.Sprintf("%s:%d", context.ToriiBaseAddress, toriiPort)
	path := fmt.Sprintf("v1alpha/instance/%s/task/%s", instanceUuid, taskUuid)
	vars := map[string]interface{}{}
	return SendGet(context, address, path, vars)
}

func ListTask(context AccessContext, toriiPort int, instanceUuid string) (map[string]interface{}, error) {
	address := fmt.Sprintf("%s:%d", context.ToriiBaseAddress, toriiPort)
	path := fmt.Sprintf("v1alpha/instance/%s/task", instanceUuid)
	vars := map[string]interface{}{}
	return SendGet(context, address, path, vars)
}

func DeleteTask(context AccessContext, toriiPort int, taskUuid, instanceUuid string) (map[string]interface{}, error) {
	address := fmt.Sprintf("%s:%d", context.ToriiBaseAddress, toriiPort)
	path := fmt.Sprintf("v1alpha/instance/%s/task/%s", instanceUuid, taskUuid)
	vars := map[string]interface{}{}
	return SendDelete(context, address, path, vars)
}

func AbortTask(context AccessContext, toriiPort int, taskUuid, instanceUuid string) (map[string]interface{}, error) {
	address := fmt.Sprintf("%s:%d", context.ToriiBaseAddress, toriiPort)
	path := fmt.Sprintf("v1alpha/instance/%s/task/%s/abort", instanceUuid, taskUuid)
	vars := map[string]interface{}{}
	return SendPut(context, address, path, vars)
}
